README.md
# unicode-charnames
[![PyPI Version](https://img.shields.io/pypi/v/unicode-charnames.svg)](https://pypi.python.org/pypi/unicode-charnames) [![PyPI License](https://img.shields.io/pypi/l/unicode-charnames.svg)](https://pypi.python.org/pypi/unicode-charnames)

Unicode characters have names that serve as unique identifiers for each character. The character names in the Unicode Standard are identical to those of ISO/IEC 10646.

The unicode-charnames package performs searches for Unicode character names or code point labels by Unicode character, and searches for Unicode code points by character names. It also performs substring searches in Unicode character names. This package supports version 14.0 of the Unicode Standard (144,697 characters).

The generic term “character name” refers to the Unicode character “Name” property value for an encoded Unicode character. For code points that do not have character names (unassigned, reserved code points and other special code point types), the Unicode Standard uses constructed Unicode code point labels, displayed between angle brackets, to stand in for character names.

### Installation
```shell
pip install unicode-charnames
```

### Features
The library provides:

* A function to get the character name (the normative character property “Name”) or the code point label (for characters that do not have character names) of a single Unicode character.
* A function to get the code point value (in the usual 4- to 6-digit hexadecimal format) corresponding to a Unicode character name; the search is case-sensitive and requires exact string match.
* A function to search characters by character name; the search is case-insensitive but requires exact substring match.

### Example usage
```python
    from unicode_charnames import charname, codepoint, search_charnames

    # charname

    for char in '龠💓\u00E5\u0002':
        print(charname(char))
        # Output:
        # CJK UNIFIED IDEOGRAPH-9FA0
        # BEATING HEART
        # LATIN SMALL LETTER A WITH RING ABOVE
        # <control-0002>

    # codepoint

    names = [
        'LATIN CAPITAL LETTER E WITH ACUTE',
        'SQUARE ERA NAME REIWA',
        'SUPERCALIFRAGILISTICEXPIALIDOCIOUS'
    ]

    for name in names:
        print(codepoint(name))
        # Output:
        # 00C9
        # 32FF
        # None

    # search_charnames

    for x in search_charnames('break'):
        print('\t'.join(x))
        # Output:
        # 00A0    NO-BREAK SPACE
        # 2011    NON-BREAKING HYPHEN
        # 202F    NARROW NO-BREAK SPACE
        # 4DEA    HEXAGRAM FOR BREAKTHROUGH
        # FEFF    ZERO WIDTH NO-BREAK SPACE
```

### References
* https://www.unicode.org/versions/Unicode14.0.0/ch04.pdf#G2082
* https://www.unicode.org/Public/14.0.0/ucd/UnicodeData.txt
* https://www.unicode.org/Public/14.0.0/ucd/extracted/DerivedName.txt

### License
unicode-charnames is released under an MIT license. The full text of the license is available [here](https://github.com/mlodewijck/unicode_charnames/blob/master/LICENSE).

The Unicode Standard “DerivedName.txt” file is licensed under the Unicode License Agreement for Data Files and Software. Please consult the [UNICODE, INC. LICENSE AGREEMENT](https://www.unicode.org/license.html) prior to use.

### Changes
* [CHANGELOG](https://github.com/mlodewijck/unicode_charnames/blob/master/CHANGELOG.md)


CHANGELOG.md
## Unicode-charnames Changelog

#### [14.0.0]
- Upgrade to Unicode 14.0.0.
- Code improvements.

#### [13.0.0]
- Upgrade to Unicode 13.0.0.
- Code improvements.

#### [12.1.0.post1]
- Code cleanup.

#### [12.1.0]
- Initial commit.


unicode_charnames/tests/test_charnames.py
# -*- coding: utf-8 -*-

import unittest

from unicode_charnames import (
    charname,
    codepoint,
    search_charnames,
    unidata_version
)


class TestSuite(unittest.TestCase):

    def test_unidata_version(self):
        self.assertEqual(unidata_version(), '12.1.0')

    def test_charname(self):
        self.assertEqual(
            charname('É'),
            'LATIN CAPITAL LETTER E WITH ACUTE')
        self.assertEqual(
            charname('\u00C9'),
            'LATIN CAPITAL LETTER E WITH ACUTE')
        self.assertEqual(
            charname(chr(0x00C9)),
            'LATIN CAPITAL LETTER E WITH ACUTE')
        self.assertEqual(
            charname('\u32FF'),
            'SQUARE ERA NAME REIWA')

        self.assertEqual(
            charname('\u3400'),
            'CJK UNIFIED IDEOGRAPH-3400')
        self.assertEqual(
            charname('\uF900'),
            'CJK COMPATIBILITY IDEOGRAPH-F900')
        self.assertEqual(
            charname('\U00017000'),
            'TANGUT IDEOGRAPH-17000')
        self.assertEqual(
            charname('\U0001B170'),
            'NUSHU CHARACTER-1B170')

        self.assertEqual(
            charname('\u0000'),
            '<control-0000>')
        self.assertEqual(
            charname('\uF8FF'),
            '<private-use-F8FF>')
        self.assertEqual(
            charname('\uD800'),
            '<surrogate-D800>')
        self.assertEqual(
            charname('\U0010FFFF'),
            '<noncharacter-10FFFF>')

    def test_codepoint(self):
        self.assertEqual(
            codepoint('LATIN CAPITAL LETTER E WITH ACUTE'),
            '00C9')
        self.assertEqual(
            codepoint('SQUARE ERA NAME REIWA'),
            '32FF')
        self.assertEqual(
            codepoint('SUPERCALIFRAGILISTICEXPIALIDOCIOUS'),
            None)

    def test_search_charnames(self):
        lst = [
            ('32FF', 'SQUARE ERA NAME REIWA'),
            ('337B', 'SQUARE ERA NAME HEISEI'),
            ('337C', 'SQUARE ERA NAME SYOUWA'),
            ('337D', 'SQUARE ERA NAME TAISYOU'),
            ('337E', 'SQUARE ERA NAME MEIZI')
        ]

        self.assertEqual(
            search_charnames('ERA NAME'),
            lst)
        self.assertEqual(
            search_charnames('ERA NAME'),
            search_charnames('era name'))

if __name__ == '__main__':
    unittest.main()


unicode_charnames/tests/test_unicode_charnames.py
import unittest

from unicode_charnames import (
    charname,
    codepoint,
    search_charnames,
    UNICODE_VERSION,
    UCD_VERSION,
)


class TestSuite(unittest.TestCase):

    def test_UNICODE_VERSION(self):
        self.assertEqual(UNICODE_VERSION, "14.0.0")

    def test_UCD_VERSION(self):
        self.assertEqual(UCD_VERSION, "14.0.0")

    def test_charname(self):
        expected = "LATIN CAPITAL LETTER E WITH ACUTE"
        self.assertEqual(charname("É"), expected)
        self.assertEqual(charname("\u00C9"), expected)
        self.assertEqual(charname(chr(0xC9)),expected)

        self.assertEqual(charname("\u3400"), "CJK UNIFIED IDEOGRAPH-3400")
        self.assertEqual(charname("\U0003134A"), "CJK UNIFIED IDEOGRAPH-3134A")
        self.assertEqual(charname("\uF900"), "CJK COMPATIBILITY IDEOGRAPH-F900")
        self.assertEqual(charname("\U00017000"), "TANGUT IDEOGRAPH-17000")
        self.assertEqual(charname("\U0001B170"), "NUSHU CHARACTER-1B170")
        self.assertEqual(charname("\U00018CD5"), "KHITAN SMALL SCRIPT CHARACTER-18CD5")

        self.assertEqual(charname("\u0000"), "<control-0000>")
        self.assertEqual(charname("\uF8FF"), "<private-use-F8FF>")
        self.assertEqual(charname("\uD800"), "<surrogate-D800>")
        self.assertEqual(charname("\U0010FFFF"), "<noncharacter-10FFFF>")

    def test_codepoint(self):
        self.assertEqual(codepoint("SQUARE ERA NAME REIWA"), "32FF")
        self.assertEqual(codepoint("BUBBLE TEA"), "1F9CB")
        self.assertIsNone(codepoint("SUPERCALIFRAGILISTICEXPIALIDOCIOUS"))

    def test_search_charnames(self):
        expected = [
            ("32FF", "SQUARE ERA NAME REIWA"),
            ("337B", "SQUARE ERA NAME HEISEI"),
            ("337C", "SQUARE ERA NAME SYOUWA"),
            ("337D", "SQUARE ERA NAME TAISYOU"),
            ("337E", "SQUARE ERA NAME MEIZI"),
        ]

        self.assertEqual(list(search_charnames("ERA NAME")), expected)
        self.assertEqual(list(search_charnames("era name")), expected)

        expected = [
            ("26B9", "SEXTILE"),
            ("26BA", "SEMISEXTILE"),
        ]

        self.assertEqual(list(search_charnames("sextile")), expected)


if __name__ == "__main__":
    unittest.main()


unicode_charnames/charnames.py
"""unicode_charnames.charnames"""

# Standard library imports
import os.path as _p
import re

# Local imports
from unicode_charnames import UCD_VERSION

_UNICODE_FILE = "DerivedName.txt"


def _make_names_dict():
    character_names = {}

    def convert(cp_range, prefix):
        start, end = cp_range.split("..")
        character_names.update({
            cp: f"{prefix}{cp:04X}"
            for cp in range(int(start, 16), int(end, 16) + 1)
        })

    with open(_p.join(_p.dirname(__file__), _UNICODE_FILE), "r") as f:
        version = re.match(
            r"^#\s*DerivedName-(.+)\.txt.*$", f.readline()
        ).group(1)
        if version != UCD_VERSION:
            raise SystemExit(f"\n{__package__}: wrong UCD version number.")
        for line in f:
            line = line.strip()
            if line and not line.startswith("#"):
                cp, name = [x.strip() for x in line.split(";")]
                if ".." in cp:
                    convert(cp, name[:-1])
                else:
                    character_names[int(cp, 16)] = name

    return character_names


# Normative Unicode character names
_CHARACTER_NAMES = _make_names_dict()

# Inverted character names dictionary
_CHARACTER_NAMES_INV = {v: k for k, v in _CHARACTER_NAMES.items()}

# assert len(_CHARACTER_NAMES_INV) == len(_CHARACTER_NAMES)  # checked


_control = [
    *range(0x000000, 0x00001F + 1),
    *range(0x00007F, 0x00009F + 1),
]

_private_use = [
    *range(0x00E000, 0x00F8FF + 1),
    *range(0x0F0000, 0x0FFFFD + 1),
    *range(0x100000, 0x10FFFD + 1),
]

_surrogate = [
    *range(0x00D800, 0x00DFFF + 1)
]

_noncharacter = [
    *range(0x00FDD0, 0x00FDEF + 1),
    *range(0x00FFFE, 0x00FFFF + 1),
    *range(0x01FFFE, 0x01FFFF + 1),
    *range(0x02FFFE, 0x02FFFF + 1),
    *range(0x03FFFE, 0x03FFFF + 1),
    *range(0x04FFFE, 0x04FFFF + 1),
    *range(0x05FFFE, 0x05FFFF + 1),
    *range(0x06FFFE, 0x06FFFF + 1),
    *range(0x07FFFE, 0x07FFFF + 1),
    *range(0x08FFFE, 0x08FFFF + 1),
    *range(0x09FFFE, 0x09FFFF + 1),
    *range(0x0AFFFE, 0x0AFFFF + 1),
    *range(0x0BFFFE, 0x0BFFFF + 1),
    *range(0x0CFFFE, 0x0CFFFF + 1),
    *range(0x0DFFFE, 0x0DFFFF + 1),
    *range(0x0EFFFE, 0x0EFFFF + 1),
    *range(0x0FFFFE, 0x0FFFFF + 1),
    *range(0x10FFFE, 0x10FFFF + 1),
]

_LABELS = {
    frozenset(_control)      : "control",
    frozenset(_private_use)  : "private-use",
    frozenset(_surrogate)    : "surrogate",
    frozenset(_noncharacter) : "noncharacter",
}

# Code points assigned to an abstract character
# print(f"{len(_CHARACTER_NAMES):,}")      # 144,697
# print(f"{len(_CHARACTER_NAMES_INV):,}")  # 144,697

# Code points with a normative function
# print(f"{len(_control):,}")              #      65
# print(f"{len(_private_use):,}")          # 137,468
# print(f"{len(_surrogate):,}")            #   2,048
# print(f"{len(_noncharacter):,}")         #      66


def charname(char):
    """Return the Unicode name or the code point label of a single
    Unicode character.

    >>> charname("A")
    'LATIN CAPITAL LETTER A'

    >>> charname("龠")
    'CJK UNIFIED IDEOGRAPH-9FA0'

    >>> charname("\U00012F90")
    'CYPRO-MINOAN SIGN CM001'

    >>> charname("\uA7F4")
    'MODIFIER LETTER CAPITAL Q'
    """
    char = ord(char)

    if char in _CHARACTER_NAMES:
        return _CHARACTER_NAMES[char]

    for set_, label in _LABELS.items():
        if char in set_:
            _CHARACTER_NAMES[char] = f"<{label}-{char:04X}>"
            break
    else:
        _CHARACTER_NAMES[char] = f"<reserved-{char:04X}>"

    return _CHARACTER_NAMES[char]


def codepoint(name):
    """Return the Unicode code point (in the usual 4- to 6-digit
    hexadecimal format) corresponding to a Unicode character name.
    The search is case-sensitive and requires exact string match.

    >>> codepoint("LATIN CAPITAL LETTER E WITH ACUTE")
    '00C9'

    >>> codepoint("MODIFIER LETTER CAPITAL Q")
    'A7F4'

    >>> codepoint('BUBBLE TEA')
    '1F9CB'

    >>> codepoint("SUPERCALIFRAGILISTICEXPIALIDOCIOUS")
    >>>
    """
    if name in _CHARACTER_NAMES_INV:
        return f"{_CHARACTER_NAMES_INV[name]:04X}"
    return None


def search_charnames(substr):
    """Search characters by character name. Input value: word or phrase
    from the normative Unicode character name. The search is case-
    insensitive but requires exact substring match.

    >>> for each in search_charnames("sextile"): print(each)
    ...
    ('26B9', 'SEXTILE')
    ('26BA', 'SEMISEXTILE')

    >>> for each in search_charnames("French"): print(each)
    ...
    ('20A3', 'FRENCH FRANC SIGN')
    ('1F35F', 'FRENCH FRIES')
    """
    for name, codepoint in _CHARACTER_NAMES_INV.items():
        if substr.upper() in name:
            yield f"{codepoint:04X}", name


if __name__ == "__main__":
    import doctest
    doctest.testmod()


unicode_charnames/core.py
# -*- coding: utf-8 -*-

import os.path
import sys


UNIDATA_VERSION = '12.1.0'
FILE = 'DerivedName.txt'
PATH = os.path.join(os.path.dirname(os.path.abspath(__file__)), FILE)

names = {}

with open(PATH, 'r', encoding='utf-8') as f:
    for line in f:
        line = line.strip()
        if not line or line.startswith('#'):
            continue
        cp, name = [x.strip() for x in line.split(';')]
        if '..' in cp:
            start, stop = cp.split('..')
            for each in range(int(start, 16), int(stop, 16)+1):
                names[each] = name[:-1] + format(each, '04X')
        else:
            names[int(cp, 16)] = name

names_inv = {v: k for k, v in names.items()}

control = set()
private_use = set()
surrogate = set()
noncharacter = set()

control.update(
    range(0x000000, 0x00001F+1),
    range(0x00007F, 0x00009F+1)
)
private_use.update(
    range(0x00E000, 0x00F8FF+1),
    range(0x0F0000, 0x0FFFFD+1),
    range(0x100000, 0x10FFFD+1)
)
surrogate.update(
    range(0x00D800, 0x00DFFF+1)
)
noncharacter.update(
    range(0x00FDD0, 0x00FDEF+1),
    range(0x00FFFE, 0x00FFFF+1),
    range(0x01FFFE, 0x01FFFF+1),
    range(0x02FFFE, 0x02FFFF+1),
    range(0x03FFFE, 0x03FFFF+1),
    range(0x04FFFE, 0x04FFFF+1),
    range(0x05FFFE, 0x05FFFF+1),
    range(0x06FFFE, 0x06FFFF+1),
    range(0x07FFFE, 0x07FFFF+1),
    range(0x08FFFE, 0x08FFFF+1),
    range(0x09FFFE, 0x09FFFF+1),
    range(0x0AFFFE, 0x0AFFFF+1),
    range(0x0BFFFE, 0x0BFFFF+1),
    range(0x0CFFFE, 0x0CFFFF+1),
    range(0x0DFFFE, 0x0DFFFF+1),
    range(0x0EFFFE, 0x0EFFFF+1),
    range(0x0FFFFE, 0x0FFFFF+1),
    range(0x10FFFE, 0x10FFFF+1)
)

def charname(char):
    """Return the Unicode name or the code point label of a single
    Unicode character.

    >>> charname('龠')
    'CJK UNIFIED IDEOGRAPH-9FA0'
    >>> charname('\U0001B170')
    'NUSHU CHARACTER-1B170'
    >>> charname('\u32FF')
    'SQUARE ERA NAME REIWA'
    """
    char = ord(char)

    name = names.get(char, None)
    if name is not None:
        return name

    # Constructed Unicode code point labels:
    if char in control:
        label = '<control-'
    elif char in private_use:
        label = '<private-use-'
    elif char in surrogate:
        label = '<surrogate-'
    elif char in noncharacter:
        label = '<noncharacter-'
    else:
        label = '<reserved-'

    name = names[char] = label + format(char, '04X') + '>'
    return name

def codepoint(name):
    """Return the Unicode code point (in the usual 4- to 6-digit
    hexadecimal format) corresponding to a Unicode character name. The
    search is case-sensitive and requires exact string match.

    >>> codepoint('LATIN CAPITAL LETTER E WITH ACUTE')
    '00C9'
    >>> codepoint('SQUARE ERA NAME REIWA')
    '32FF'
    >>> codepoint('SUPERCALIFRAGILISTICEXPIALIDOCIOUS')
    >>> 
    """
    if name in names_inv:
        return format(names_inv[name], '04X')
    return None

def search_charnames(substr):
    """Search characters by character name. Input value: word or phrase
    from the normative Unicode character name. The search is case-
    insensitive but requires exact substring match.

    >>> type(search_charnames('ERA NAME'))
    <class 'list'>
    >>> type(search_charnames('ERA NAME')[0])
    <class 'tuple'>
    >>> for each in search_charnames('era name'):
    ...     print(each)
    ...
    ('32FF', 'SQUARE ERA NAME REIWA')
    ('337B', 'SQUARE ERA NAME HEISEI')
    ('337C', 'SQUARE ERA NAME SYOUWA')
    ('337D', 'SQUARE ERA NAME TAISYOU')
    ('337E', 'SQUARE ERA NAME MEIZI')
    """
    res = []
    for name in names.values():
        if substr.upper() in name:
            res.append((codepoint(name), name))
    return res

def unidata_version():
    return UNIDATA_VERSION


if __name__ == '__main__':

    # Package version:
    from release import __version__
    print(__version__)

    # Unicode version:
    print(unidata_version())

    # Code points assigned to an abstract character:
    print(len(names))         # 137929

    # Code points with a normative function:
    print(len(control))       #     65
    print(len(private_use))   # 137468
    print(len(surrogate))     #   2048
    print(len(noncharacter))  #     66

    # Docstrings testing:
#    import doctest
#    doctest.testmod()
